[WIP] Isometric grid

Liebranca opened the pull request:
Implements an isometric grid and 'Snap To' for it. Part of #720.

The projection angle, so to speak, depends entirely on grid size, and so multiple styles are supported from the get-go:

![cubes](https://github.com/user-attachments/assets/6d2784f9-a05f-4044-a126-82244d419860)


But although this is very flexible, it has it's share of issues: not every line angle will give an equally good-looking result for pixel art.


A summary of what's in this PR:

- Drawing of the isometric grid itself, plus a pixel-precise version when zoomed in.

- Added the `Grid::Type` enum to identify which type of grid a sprite is using, plus a drop-down element in grid settings to select between orthogonal (default) or isometric, as well as handling the persistence of this setting.

- Implementation of 'Snap To' for isometric, with a few adjustments for the line tool so that it can be used for drawing.

- When using the line tool and with 'Snap To' enabled, vertical lines can be selectively offset one pixel to the left or right by holding or releasing the 'Angle Snap' key. I added this merely so I could draw tileable cubes quicker as I did a lot of that during testing. I left it there just in case.


And a few things that are still pending or need further polish:

- Offering better control over the pixel-precise grid, for instance, giving an option to toggle off shared edges between cells, or allowing setting specific border width and height.

- Following from the offsetting of vertical lines, I would give the option to set a specific offset value for it, as this comes into play when selecting or moving isometric tiles, and it's not the same offsets for every style.

- I applied the aforementioned line adjustment to the Freehand controller, but it's not a 100% correct, as the end point of one line doesn't always match the start point of the next.

- Because odd sizes are generally problematic, I've set them to be (in a sense) silently ignored via sharing of edges to make them evenly divisible by two, but this doesn't always work as intended.

- Due to how lines intertwine at certain angles, the direction of a stroke matters when drawing (`line(A, B) != line(B, A)`), which may be a bit confusing. I've remedied it for a handful of cases when using the line tool with 'Snap to Grid' enabled, but this is not a complete fix.

- Similarly, length of a stroke is also a factor for `TwoPointsController`, also dependant on line angle.

- There are no adjustments for rectangles and ellipses, only lines.

- With some odd grid sizes, like 30x15 or 18x9, the snap point will be off by one pixel.

Liebranca committed:
Implement isometric grid

Liebranca committed:
Implement snap to isometric grid

dacap commented:
@Liebranca thanks a lot for all this work! :pray: I'll try to test this step by step, starting tomorrow. It'd be great to integrate a first iteration of the feature soon in the beta.

Liebranca committed:
Add pixel-precise isometric grid

Liebranca committed:
Add persistence and grid offset for isometric

Liebranca committed:
Add pixel-precise grid for any grid size

Liebranca committed:
Adjust line tool for isometric 'snap to'

Liebranca committed:
Fix isometric 'snap to' for odd grid sizes

Liebranca committed:
Revise implementation of isometric 'Snap To'

dacap commented on src/app/tools/controllers.h:
```diff
@@ -17,6 +17,94 @@ namespace app { namespace tools {
+  Stroke::Pt& b = stroke[stroke.size() - 1];
+
+  // Get function invoked by line tool
+  bool lineTool = (string_id_to_brush_type(loop->getTool()->getId()) == kLineBrushType);
```
Do we need to check the line tool or the line brush type? because here we are passing the tool ID to a function that receives a brush type (and I think it works only by chance because `"line"` is a brush type and a tool ID).

Probably you are looking for something like:
```suggestion
  const bool lineTool = (loop->getTool()->getId() == WellKnownTools::Line);
```
But you will need to add:
```c++
const char* WellKnownTools::Line = "line";
```
in `tool_box.cpp` and `.h`.

dacap commented on src/app/ui/editor/editor.cpp:
```diff
@@ -1145,21 +1148,156 @@ void Editor::drawGrid(Graphics* g,
-  int x2 = spriteBounds.x + spriteBounds.w;
-  int y2 = spriteBounds.y + spriteBounds.h;
+  // Orthogonal grid
+  if (isPixelGrid || getSite().sprite()->gridType() == doc::Grid::Type::Orthogonal) {
```
We have to avoid calling `getSite()` as much as possible (the site collects a lot of information that might be unnecessary):
```suggestion
  if (isPixelGrid || m_sprite->gridType() == doc::Grid::Type::Orthogonal) {
```

dacap commented on src/app/tools/controllers.h:
```diff
@@ -17,6 +17,94 @@ namespace app { namespace tools {
+      // for freehand strokes would require changes to intertwiners,
+      // not just the freehand controller itself.
+      if (lineTool && vto.x < 0 && a.x % (grid.w - !line.evenWidth)) {
+        auto tmp = a;
+        a = b;
+        b = tmp;
```
You can use `std::swap()` (needs `<utility>` probably)
```suggestion
        std::swap(a, b);
```

dacap commented on src/doc/grid.cpp:
```diff
@@ -177,4 +179,60 @@ std::vector<gfx::Point> Grid::tilesInCanvasRegion(const gfx::Region& rgn) const
+
+  // TODO: add 'share edges' checkbox to UI.
+  // For testing the option, set this 'false' to 'true'
+  shareEdges = !(false && !squareRatio && evenHeight);
```
I'm thinking if the shareEdges could be an argument of this function, or another type of grid `IsometricSharedEdges` or something like that (and passing the grid type to `IsometricGuide` ctor).

dacap reviewed (changes requested):
I was finally able to give a try and test a little this implementation. So far so good 🚀 There are some final details to merge the PR, and after that we can fix any future issue/bug getting some user feedback.

Liebranca committed:
Fix UI/command bugs related to isometric grid

Liebranca committed:
Use doc::Grid for first argument of snap_to_grid

Includes changes to doc::Grid::Type to enable shared/thick edges
for isometric grid, as well as minor fixes for isometric 'Snap To'

Liebranca commented:
Hi, @dacap! I've separated the changes into two commits as the modification to `snap_to_grid()` meant I had to touch a few additional files, hopefully splitting it up makes it easier to amend/revert things later on if we need to.

A quick note about `GridSettingsCommand`/`OptionsWindow`, I added two helper functions to convert the string to a `Grid::Type` and viceversa, I'm not really sure which file it'd be best to put them in. For now I've placed these functions in `app/cmd/set_grid_type.h` as both classes already include this header. I noticed a few other minor things I initially missed regarding setting/getting the default type so I've addressed them as well.

I also added a few methods to `doc::Grid` to try and take advantage of the object now being used more directly, and also a new type of grid to act as toggle for the `shareEdges` option. I've updated the strings to convey the difference between the two types of isometric grid, however the new type (I've called it "thick edges" for now) doesn't work if the grid's height is uneven, and in such cases the two types are effectively the same, we may want to communicate this somehow (could there be a tooltip perhaps?).

Finally, I noticed a couple minor problems with snapping, the more important being in `MovingCelState` as the `calcPivot()` method didn't account for the grid being isometric (I think I was missing patch #4646 when I first wrote this), so I added a small fix for these as well.

dacap committed:
Merge branch 'beta' into isometric-grid

dacap commented:
Rebased in https://github.com/aseprite/aseprite/pull/5793
